### PR TITLE
Mobile broadband support

### LINKED ISSUE #2: Cannot configure Mobile Broadband devices [$100]
uname -a:

Linux darek-LIFEBOOK-E744 4.4.0-31-generic #50-Ubuntu SMP Wed Jul 13 00:07:12 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

dmesg:

[    2.860995] usb 3-11: config 1 has an invalid interface number: 8 but max is 3
[    2.860999] usb 3-11: config 1 has no interface number 1
[    2.861412] usb 3-11: config 2 has an invalid interface number: 12 but max is 1
[    2.861415] usb 3-11: config 2 has an invalid interface number: 13 but max is 1
[    2.861416] usb 3-11: config 2 has an invalid interface number: 13 but max is 1
[    2.861418] usb 3-11: config 2 has no interface number 0
[    2.861419] usb 3-11: config 2 has no interface number 1
[    2.862365] usb 3-11: New USB device found, idVendor=1199, idProduct=9041
[    2.862368] usb 3-11: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[    2.862369] usb 3-11: Product: EM8805

lsusb -v:

Bus 003 Device 005: ID 1199:9041 Sierra Wireless, Inc. 
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 (Defined at Interface level)
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1199 Sierra Wireless, Inc.
  idProduct          0x9041 
  bcdDevice            0.06
  iManufacturer           1 
  iProduct                2 
  iSerial                 3 
  bNumConfigurations      2
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength          160
    bNumInterfaces          4
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xe0
      Self Powered
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           2
      bInterfaceClass       255 Vendor Specific Class
      bInterfaceSubClass    255 Vendor Specific Subclass
      bInterfaceProtocol    255 Vendor Specific Protocol
      iInterface              0 
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0200  1x 512 bytes
        bInterval               0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x01  EP 1 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0200  1x 512 bytes
        bInterval               0
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           3
      bInterfaceClass       255 Vendor Specific Class
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              0 
      ** UNRECOGNIZED:  05 24 00 10 01
      ** UNRECOGNIZED:  05 24 01 00 00
      ** UNRECOGNIZED:  04 24 02 02
      ** UNRECOGNIZED:  05 24 06 00 00
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x000c  1x 12 bytes
        bInterval               9
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0200  1x 512 bytes
        bInterval               0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x02  EP 2 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0200  1x 512 bytes
        bInterval               0
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        3
      bAlternateSetting       0
      bNumEndpoints           3
      bInterfaceClass       255 Vendor Specific Class
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              0 
      ** UNRECOGNIZED:  05 24 00 10 01
      ** UNRECOGNIZED:  05 24 01 00 00
      ** UNRECOGNIZED:  04 24 02 02
      ** UNRECOGNIZED:  05 24 06 00 00
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x85  EP 5 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x000c  1x 12 bytes
        bInterval               9
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x84  EP 4 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0200  1x 512 bytes
        bInterval               0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x03  EP 3 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0200  1x 512 bytes
        bInterval               0
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        8
      bAlternateSetting       0
      bNumEndpoints           3
      bInterfaceClass       255 Vendor Specific Class
      bInterfaceSubClass    255 Vendor Specific Subclass
      bInterfaceProtocol    255 Vendor Specific Protocol
      iInterface              0 
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x87  EP 7 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x000a  1x 10 bytes
        bInterval               9
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x86  EP 6 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0200  1x 512 bytes
        bInterval               0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x04  EP 4 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0200  1x 512 bytes
        bInterval               0
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength           95
    bNumInterfaces          2
    bConfigurationValue     2
    iConfiguration          0 
    bmAttributes         0xe0
      Self Powered
      Remote Wakeup
    MaxPower              500mA
    Interface Association:
      bLength                 8
      bDescriptorType        11
      bFirstInterface        12
      bInterfaceCount         2
      bFunctionClass          2 Communications
      bFunctionSubClass      14 
      bFunctionProtocol       0 
      iFunction               0 
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber       12
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         2 Communications
      bInterfaceSubClass     14 
      bInterfaceProtocol      0 
      iInterface              0 
      CDC Header:
        bcdCDC               1.10
      CDC Union:
        bMasterInterface        12
        bSlaveInterface         13 
      CDC MBIM:
        bcdMBIMVersion       1.00
        wMaxControlMessage   4096
        bNumberFilters       32
        bMaxFilterSize       128
        wMaxSegmentSize      1500
        bmNetworkCapabilities 0x20
          8-byte ntb input size
      UNRECOGNIZED CDC:  08 24 1c 00 01 40 dc 05
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               9
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber       13
      bAlternateSetting       0
      bNumEndpoints           0
      bInterfaceClass        10 CDC Data
      bInterfaceSubClass      0 Unused
      bInterfaceProtocol      2 
      iInterface              0 
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber       13
      bAlternateSetting       1
      bNumEndpoints           2
      bInterfaceClass        10 CDC Data
      bInterfaceSubClass      0 Unused
      bInterfaceProtocol      2 
      iInterface              0 
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0200  1x 512 bytes
        bInterval               0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x01  EP 1 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0200  1x 512 bytes
        bInterval               0

usb-devices:

T:  Bus=03 Lev=01 Prnt=01 Port=10 Cnt=02 Dev#=  6 Spd=480 MxCh= 0
D:  Ver= 2.00 Cls=00(>ifc ) Sub=00 Prot=00 MxPS=64 #Cfgs=  2
P:  Vendor=1199 ProdID=9041 Rev=00.06
S:  Manufacturer=Sierra Wireless, Incorporated
S:  Product=EM8805
C:  #Ifs= 2 Cfg#= 2 Atr=e0 MxPwr=500mA
/usr/bin/usb-devices: line 79: printf: c: invalid number
I:  If#= 0 Alt= 0 #EPs= 1 Cls=02(commc) Sub=0e Prot=00 Driver=cdc_mbim
/usr/bin/usb-devices: line 79: printf: d: invalid number
I:  If#= 0 Alt= 1 #EPs= 2 Cls=0a(data ) Sub=00 Prot=02 Driver=cdc_mbim

Launchpad Details: [#LP1609632](https://bugs.launchpad.net/bugs/1609632) Dariusz Brzezinski - 2016-08-04 04:30:01 +0000

### PR DESCRIPTION
**Pull request by davidmhewitt**
Fixes #2.

Provides an interface that looks like this for each mobile broadband device:
image

Currently, the NetworkManager GUI is used to configure the connection initially with the buttons at the bottom of the view, similar to how wired devices are configured. This follows the style of the other views and there is an open bug about that in #8.

### CODE REVIEW

**Comment by matzipan** [comment]
1. If I forcibly remove the device while the panel is open, I see this:
```
[FATAL 22:37:01.622970] [Gtk] gtk_list_box_row_get_index: assertion 'GTK_IS_LIST_BOX_ROW (row)' failed
```
2. If I then add the device while the panel is open, the device doesn't get added until i close and reopen the plug.
3. In the video [linked](https://photos.app.goo.gl/Egsuyu4wTm05CAmP2) please confirm that the "Unknown" fields are caused by the driver and that the toggling issues do not come from this branch.
4. In the indicator, the device appears as disabled, whereas in switchboard the device appears as enabled with a label of "Failed" underneath.

**Comment by davidmhewitt** [comment]
@matzipan Thank you for the testing.

1. That error message is unrelated to this branch. I get the same message unplugging a USB ethernet adapter, so I think that's outside the scope of this PR.
2. When plugging a mobile broadband device back in again on my computer, it does re-appear in the list. Could you check to see whether it appears back in DBus after re-plugging? You should see an extra device be added to `/org/freedesktop/NetworkManager` (for example `/org/freedesktop/NetworkManager/Devices/10`).
3+4. I assume you're testing with a device that does not have a valid SIM card or is considered "Failed" by NetworkManager for some other reason? I've pushed a commit that disables the device in switchboard as well as the indicator in these cases.

**Comment by matzipan** [comment]
2. It does not appear
3. Could it ever reach a failed state as a consequence of a setting? Does this prevent the user from changing that setting?

**Comment by davidmhewitt** [comment]
@matzipan Thanks again for taking another look.

If the device isn't appearing to NetworkManager, it won't appear in the plug either since that's where the device list is coming from. I suspect there's something about the driver used with your device that means it doesn't register on the fly but opening the plug again causes NetworkManager to re-scan for devices.

As for the failed state, that state is described in the network manager docs as being issues with drivers or hardware. Nevertheless, the "edit connections" button is still available even while in the failed state to allow you to change any settings for the device. The connection simply can't be enabled while in the failed state.

File: src/common/Widgets/AbstractModemInterface.vala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    public override void update_name (int count) {
+        var name = device.get_description ();
+        if (count > 1) {
+            display_title = _("Mobile Broadband: %s").printf (name);
+        } else {
+            display_title = _("Mobile Broadband");
+        }
+
+        if (device is NM.DeviceModem) {
+            var capabilities = (device as NM.DeviceModem).get_current_capabilities ();
```
Comment (donadigo) [inline src/common/Widgets/AbstractModemInterface.vala]: Do an explicit `((NM.DeviceModem)device)` cast here.

File: src/common/Widgets/AbstractModemInterface.vala [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            if (NM.DeviceModemCapabilities.POTS in capabilities) {
```
Comment (donadigo) [inline src/common/Widgets/AbstractModemInterface.vala]: What if there's more of these?

File: src/common/Widgets/AbstractModemInterface.vala [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+ *
+ * You should have received a copy of the GNU Library General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: David Hewitt <davidmhewitt@gmail.com>
+ */
+
+public abstract class Network.AbstractModemInterface : Network.WidgetNMInterface {
+    public override void update_name (int count) {
+        var name = device.get_description ();
```
Comment (donadigo) [inline src/common/Widgets/AbstractModemInterface.vala]: Move this down to the `count > 1` block.

File: src/common/Widgets/AbstractModemInterface.vala [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        if (count > 1) {
+            display_title = _("Mobile Broadband: %s").printf (name);
+        } else {
+            display_title = _("Mobile Broadband");
+        }
+
+        if (device is NM.DeviceModem) {
```
Comment (donadigo) [inline src/common/Widgets/AbstractModemInterface.vala]: Move this statement up and make it `else if` like this:

```
if (count > 1) {
    var name = ...
    ...
} else if (device is NM.DeviceModem) {
    ....
} else {
   ....
}
```

**Pull request merged**